text/0000-wire-reform.md

```diff
@@ -0,0 +1,146 @@
+```
+
+_Notes:_
+
+* Not all environments will support or need contextualization (e.g.: preloading LDS data), the `requestContextCallback` is optional. All adapters should be able to work with and without a context provider.
+
+## Interop
+
+If you have an adapter, you should be able to use it with any component system, not just LWC. This is an example of how to use this with React:
+
+```js
+// shared adapter
+import { MyWireAdapter } from 'some-module';
+
+class Foo extends React.Component {
+    constructor(props) {
+        super(props);
+        // The wire adapter instance is bound to the host object via the callback for data
+        this.adapter = new MyWireAdapter((data) => {
+            this.setState(() => {
+                // stream of data from wire adapter to be used to update the component's state
+                return data;
+            });
+        });
+        // calling for the initial update of the config since componentDidUpdate() is not
+        // invoked for the first time, but all props are ready.
+        this.adapter.update({ x: 1, y: this.props.valueOfY });
+    }
+
+    componentDidUpdate(prevProps) {
+        if (this.props.valueOfY !== prevProps.valueOfY) {
+            // recompute the config by extracting `this.props.valueOfY`
+            this.adapter.update({ x: 1, y: this.props.valueOfY });
+        }
+    }
+
+    componentDidMount() {
+        this.adapter.connect();
+    }
+
+    componentWillUnmount() {
+        this.adapter.disconnect();
+    }
+
+    render() {
+        return (
+            <div>{this.state.valueProducedByMyWireAdapter}</div>
+        );
+    }
+}
+```
+
+# Adoption strategy
+
+Since there is a need to support callable adapters that behave differently depending on who uses that (wire adapter vs user invoking the function directly), we have added a simple mechanism to support such feature via `adapter` property member expression on the callable. This opens the door to transition existing adapters to the new form. E.g.: APEX adapters are all callable objects.
+
+Additionally, those callable objects can implement a forking logic based on the type of argument, if there is a desire to avoid the `adapter` property member expression.
+
+# How we teach this
+
+* For adapter consumers, nothing changes.
+* For adapter author, the wire protocol is now well defined and does not need registration, which means it is easier to reasoning about compared to the existing mechanism.
+* As for existing adapters based on `@lwc/wire-service`, they can remain the same until after they get refactored and simplified when possible.
+
+# Unresolved questions
+
+* In the current implementation, a wired field is `writable`, which means the component author can alter the value of the field at will. What should we do? a) throw on setter, b) do nothing on setter, c) preserve the current semantics. This is a breaking change if we do a) or b), while the current behavior is weird.
```

Comment:
While the semantic can be improved with a read-only field, I don't see any major issue with the current behavior.
At worst developers can override a property that is set by the `@wire` decorator, which is not harmful IMO.